proposals/dgl-operator-proposal.md

```diff
@@ -0,0 +1,284 @@
+
+### Training-wise
+
+**Default communication backend:** gloo
+
+**How does DGL implement distributed training**
+
+- data parallelism: partitioned sub-graph is like the divided data
+- the allreduce way: updating dense parameters is used `DDP` in PyTorch
+
+**How do we leverage DGL to run standalone partitioning and distributed training on Kubernetes**
+
+Prerequisite: the dataset must be included in worker `Pod` image, like [OGB](https://github.com/snap-stanford/ogb).
+
+0. Before main container of launcher `Pod` working
+    1. `partition-driver` container is finished, when partitioner `Pod` has already done the partitioning workload.
+    2. Partitioner `Pod` sleeps.
+    3. `kubectl-delivery` container is finished, when all worker `Pods` are ready.
+    4. The main container of launcher `Pod` starts running.
+1. In the main container, `dglrun` invokes a `kubectl cp` commond to copy sub-graph from partitioner `Pod` to the specific worker `Pod`. And then partitioner `Pod` will be shutdown and released.
+2. After finished graph dispatching, `dglrun` invokes distributed training on every worker `Pod`. Guide [here](https://github.com/dmlc/dgl/tree/master/examples/pytorch/graphsage/experimental#step-3-launch-distributed-jobs).
+
+**How do we leverage DGL to run distributed partitioning and training on Kubernetes**
+
+Prerequisite: the dataset must be included in worker `Pod` image, like [OGB](https://github.com/snap-stanford/ogb).
+
+0. When `kubectl-delivery` container is finished, the main container of launcher `Pod` starts running.
+1. In the main container, `dglrun` invokes a partition commond to do distributed partitioning on every worker `Pod`. Dive in [this approach](https://docs.dgl.ai/guide/distributed-preprocessing.html#distributed-partitioning).
+2. After finished distributed partitioning, `dglrun` invokes distributed training on every worker `Pod`. Guide [here](https://github.com/dmlc/dgl/tree/master/examples/pytorch/graphsage/experimental#step-3-launch-distributed-jobs).
+
+![DGL Training-wise Diagram](diagrams/dgl-operator-3.png)
+
+## Alternatives Considered
+
+### DGL Operator vs MPI Operator
+
+Currently, for distributed partitioning, DGL uses `ParMETIS` to split big graphs up and store in parallel on each worker `Pods`, which is empowered by MPI. For distributed training, DGL uses PyTorch's `DDP` in dense parameters updating, and self-developed methods in other scenarios.
+
+So in this manner, a fully-informed user, who understand any details about DGL and Operator philosophy, can hack some DGL workloads (distributed partitioning & training) on MPI Operator. But this hack approach is not a good idea, because 1) user need to revise `ipconfig` processing in DGL library to adapt MPI Operator's `hostfile`; 2) user may need involve intrusive code in partition and train script.
+
+However, there is a way to reuse MPI Operator as much as possible, MPI Operator may need,
```

Comment:
@terrytangyuan @carmark

Can you help evaluate if these can be changed in MPI-Operator to support DGL use case?

Reply:
more details at https://github.com/kubeflow/community/pull/512#issuecomment-827414177